runtime/lua/vim/treesitter/_fold.lua

```diff
@@ -347,13 +347,14 @@ function M.foldexpr(lnum)
   lnum = lnum or vim.v.lnum
   local bufnr = api.nvim_get_current_buf()
 
-  local parser = ts.get_parser(bufnr, nil, { error = false })
-  if not parser then
-    return '0'
-  end
-
   if not foldinfos[bufnr] then
     foldinfos[bufnr] = FoldInfo.new()
+    local parser = ts.get_parser(bufnr, nil, { error = false })
+
+    if not parser then
```

Comment:
nit: foldinfo for bufer without parser won't be cleaned up when the buffer is deleted.

I suggest registering a BufUnload autocmd (or similar) to clean up the foldinfo.